Dockerfile

```
@@ -66,10 +68,16 @@ RUN if [ "$INSTALL_ALL" != "false" ]; then \
 RUN . /.env && \
     if [ "$CHECKOV_VERSION" != "false" ]; then \
     ( \
-        apk add --no-cache gcc=~12 libffi-dev=~3 musl-dev=~1; \
-        [ "$CHECKOV_VERSION" = "latest" ] && pip3 install --no-cache-dir checkov \
-        || pip3 install --no-cache-dir checkov==${CHECKOV_VERSION}; \
-        apk del gcc libffi-dev musl-dev \
+        # cargo, gcc, git, musl-dev, rust and CARGO envvar required for compilation of rustworkx@0.13.2, no longer required once checkov version depends on rustworkx >0.14.0
+        # https://github.com/bridgecrewio/checkov/pull/6045
+        # gcc libffi-dev musl-dev required for compilation of cffi, until it contains musl aarch64
+        export CARGO_NET_GIT_FETCH_WITH_CLI=true && \
+        apk add --no-cache cargo=~1 gcc=~12 git=~2 libffi-dev=~3 libgcc=~12 musl-dev=~1 rust=~1 ; \
+        if [ "$CHECKOV_VERSION" = "latest" ]; \
+            then pip3 install --no-cache-dir checkov || exit 1; \
+            else pip3 install --no-cache-dir checkov==${CHECKOV_VERSION} || exit 1; \
```

`exit 1` prevents silent fail of `error: can't find Rust compiler`

---

btw, we can't just add comments inside the code about it. Probably, the best place is on L73

---

Out of interest why at `# Install pre-commit` code block (lines 20-23) `pip3` has no `|| exit 1` bits? Is ii intentional?

---

Because pre-commit don't uses rust?

I prefer somehow make structured tests working form arm64, than add exit 1 in every possible place

---

> Because pre-commit doesn't use rust?

That's kind of weird: if `pip3 install ...` fails (as in returns non-zero code) then exit the shell with non-successful code (`exit 1`). And this is implemented for `checkov` install only, which means it's okay to not exit for the same e.g. with `pre-commit` install by `pip`. Does `pip` behave differently if it fails to install `pre-commit`? 🤔 Aint't Docker's `RUN` imply somewhat `set -e`? Should we try and add `set -e` to each `RUN` to explicitly let Docker `RUN` exit when any downstream command fails within `if/else` statement? 🤔 I'm a bit lost to be honest 😲 We definitely should use consistent solution for all similar expressions for consistency

---

> When the rust compiler call fails it generates a false

exit code 0 and `false`

I dunno why, and who is responsible for such a great idea, but it's an edge case.

---

Apologies for... erm... odd suggestion, though should we wrap the shell logic of each `RUN` command into a separate shell scripts and run them instead of shell-scripting directly within `RUN` commands with all the fuss around escaping and trying to keep shell code readable with weird formatting quirks? 🤪
I do understand that this would add more layers, but... but it would ease maintenance of shell code =)
We can write up a single shell script that would handle different scenarios based on cmdline parameters (like `install_deps.sh pre-commit`, `install_deps.sh checkov`, so on) so that we `COPY` it only once.

Half past three a.m. — I might be generating odd ideas though 🤣

---

>Does pip behave differently if it fails to install pre-commit

First of all, it not fails :)
second, if pip fails to install something, it usually returns exit code 1 => docker build fails

tbh, it's looks like a bug somewhere in pip or in it deps

---

> > When the rust compiler call fails it generates a false
> 
> exit code 0 and `false`

`false` in common terms is "return non-zero code".

> I dunno why, and who is responsible for such a great idea, but it's an edge case.

The only reasonable explanation for me is "Docker `RUN` runs shell w/o `set -e`" and that is why it doesn't fail when a program returns unsuccessfully and this is why `|| exit 1` makes it fail explicitly since `exit` command forces shell to exit.

---

Regarding the installation script - why not, but it is out of the scope of this PR + not on my priority list

---

> > Does pip behave differently if it fails to install pre-commit
> 
> First of all, it not fails :)

If it does **not**, then you wouldn't be able to rely upon `||` (which is "if command om the left side returned non-successfully") or... or we definitely should **re**-RTFM =)))

> tbh, it's looks like a bug somewhere in pip or in it deps

or we're missing something very obvious 🤔

---

Prepend all `RUN` command with `set -e && ` and drop ` || exit 1` statements 🤪

---

I'll rewrite it to bash scripts. Firstly I thought it will be a part of this PR, but now I see that better split them

---

> I'll rewrite it to bash scripts.

What do you of a single script that takes args? Like `install_deps.sh pre-commit`, `install_deps.sh checkov`, `install_deps.sh pre-commit checkov ...`, or `install_deps.sh ALL`? Just to keep code in the same place and have re-usable snippets (like shell funcxtions) 🤔

---

Problem there that they are different

---

currently, I just moving it out and implementing logic as
```
COPY tools/install/ /install/

WORKDIR /bin_dir

RUN /install/pre-commit.sh
RUN /install/foo.sh
```

We can discuss it in next PR :)

---

> Problem there that they are different

Some are slightly different, whilst others use the same solutions (like extracting download URLs from GH releases API response). And also I see it odd to have a bunch of almost-the-same ten-lines shell scripts instead of one that can handle installation of all the deps one-by-one or all-at-once.
On the other hand such approach with single script would negatively impact build caching and each `RUN` layer would get rebuilt if the file is updated with a change to the installation steps of a specific dep =(
From this point if view I'd better stay with the current approach =)

> We can discuss it in next PR :)

Makes sense 🤝 (apologies that I already outlined my thoughts — this helps me imprint them in memory 😺)

---

@MaxymVlasov *do not* install distributions with pip separately because it will not take all things installed in the previous session into account when you run a new install command. Enumerate everything and let the dependency resolver know all your requirements together. Ideally, use pip-compile to produce and commit constraint files (lockfiles) and invoke it via `pip install -r direct-deps.txt -c constraint.txt`.

If you do run separate pip installs, inject a `pip check` invocation at the end to verify integrity.